### PR TITLE
[6.0][Concurrency] Fix default actor isolation issue in face of TaskExecutor preference

### PR DESCRIPTION
**Pull request by ktoso**
**Description**: The task executor API did not properly account for taking the default actor locking into account when running code on it, we just took the job and ran it without checking with the serial executor at all, which resulted in potential concurrent executions inside the actor -- violating actor isolation.

Here we change the TaskExecutor enqueue API to accept the "target" serial executor, which in practice will be either generic or a specific default actor, and coordinate with it when we perform a runSynchronously.

The SE proposal needs to be amended to showcase this new API, however without this change we are introducing races so we must do this before the API is stable.

**Scope/Impact**: Without this patch, task executors circumvent default actor locking and break actor isolation -- potentially allowing concurrent execution on given actor.
**Risk:** Low, adjusts code to take a path including locking the actor as it should have. No ABI or API impact.
**Testing**: Added tests verifying the situations this fixes
**Reviewed by**: @mikeash 

**Original PR:** https://github.com/swiftlang/swift/pull/74658
**Radar:** rdar://130337134

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/Actor.cpp [hunk 1]
```diff
@@ -1822,7 +1878,8 @@ bool DefaultActorImpl::unlock(bool forceUnlock)
 
       if (newState.isScheduled()) {
         // See ownership rule (6) in DefaultActorImpl
-        scheduleActorProcessJob(newState.getMaxPriority());
+        // FIXME: should we specify some task executor here, since otherwise we'll schedule on the global pool
```
Comment (mikeash) [inline stdlib/public/Concurrency/Actor.cpp:1881]: Does this FIXME need to be addressed in this PR, or is this more of a future todo thing?

File: stdlib/public/Concurrency/Actor.cpp [hunk 2]
```diff
@@ -2149,84 +2209,64 @@ static void swift_task_switchImpl(SWIFT_ASYNC_CONTEXT AsyncContext *resumeContex
...
     HeapObject *executor, const Metadata *selfType,
     const TaskExecutorWitnessTable *wtable);
 
 SWIFT_CC(swift)
-static void swift_task_enqueueImpl(Job *job, SerialExecutorRef executor) {
-  SWIFT_TASK_DEBUG_LOG("enqueue job %p on serial executor %p", job,
-                       executor.getIdentity());
+static void swift_task_enqueueImpl(Job *job, SerialExecutorRef serialExecutorRef) {
+#ifndef NDEBUG
+  auto _taskExecutorRef = TaskExecutorRef::undefined();
```
Comment (mikeash) [inline stdlib/public/Concurrency/Actor.cpp:2219]: The underscore made me think it's a member variable somewhere. Maybe drop that, and wrap this whole `!NDEBUG` thing in `{}` to keep the variable from leaking into the following code.

**Review by mikeash — APPROVED** [review]

File: stdlib/public/Concurrency/Actor.cpp (hunk 2 above)
Comment (ktoso) [reply stdlib/public/Concurrency/Actor.cpp:2219]: I can do that, yeap :)

File: stdlib/public/Concurrency/Actor.cpp (hunk 1 above)
Comment (ktoso) [reply stdlib/public/Concurrency/Actor.cpp:1881]: More of a future thing, though I need @rjmccall or @rokhinip to chime in how to test these cases at all as I was not able to and thus unable to address the FIXME

**Review by DougGregor — APPROVED** [review]

**Pull request merged**